keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -454,13 +460,68 @@ EndpointSliceSyncs = metrics.NewCounterVec(
 
 ```
 
+### Events
+A common point of frustration among initial users of this feature was how
+difficult it was to tell if this feature was enabled and working as intended.
+Due to the nature of this design, even when a user opts in to the `Auto` mode
+that is no guarantee that the controller logic will determine that there are
+a sufficient number of endpoints to allocate them proportionally to each zone
+in the cluster.
+
+To make this feature easier to understand and use, the EndpointSlice controller
+will publish events for a Service to describe if the feature has been enabled,
+and if not, why not.
+
+#### Logic
+
+The EndpointSlice controller will track the known state of this feature for
+each Service. When that state or the reason for it changes, the EndpointSlice
+controller will publish a new Event to reflect the updated status of this
+feature.
+
+#### Sample Events
+
+| Type | Reason | Message | 
+|-|-|-|
+| Normal | TopologyAwareRoutingEnabled | Topology Aware Routing has been enabled |
+| Normal | TopologyAwareRoutingDisabled | Topology Aware Routing configuration was removed |
+| Warning | TopologyAwareRoutingDisabled | Insufficient number of Endpoints (n), impossible to safely allocate proportionally |
+| Warning | TopologyAwareRoutingDisabled | 1 or more Endpoints do not have a Zone specified |
+| Warning | TopologyAwareRoutingDisabled | 1 or more Nodes do not have allocatable CPU specified |
+| Warning | TopologyAwareRoutingDisabled | Nodes only ready in 1 zone |
+
+#### Documentation
+
+The Topology Aware Hints documentation will be updated to describe the reason
+each of these events may have been triggered, along with steps that can be taken
+to recover from that state.
+
+#### Limitations
+
+Although the events described above should dramatically simplify the use of this
+feature, there is a tiny edge case that will not be covered. If any
+EndpointSlices for a Service do not include Hints, Kube-Proxy will not implement
+this feature. This would happen if a user created custom EndpointSlices _and_
+enabled Topology Aware Hints _and_ failed to set Hints on their custom
+EndpointSlices. This seems very unlikely, but is mentioned here for the sake of
+completeness.
+
 ### Graduation Criteria
 **Alpha:**
 - Basic functionality covered with unit tests described above.
 
 **Beta:**
 - Tests expanded to include e2e coverage described above.
 
+**GA:**
+- Feedback from real world usage shows that feature is working as intended
+- Events are triggered on each Service to provide users with clear information
+  on when the feature transitioned between enabled and disabled states.
+- Conformance tests are in place
+- Autoscaling and Scheduling SIGs have a plan to provide zone aware autoscaling
+  (and scheduling) that allows users to proportionally distribute endpoints
+  across zones.
```

Comment:
Given that you're claiming we're going to make GA in 1.26, that would imply that these plans already exist, in which case can you link to them?

Reply:
I'm not sure how this process works. I'd thought that graduation criteria had to be met prior to making the change in upstream k8s OSS, not before the KEP updates including the graduation criteria merged. I will say that it will be quite a stretch to get this to GA in 1.26, I just want to make sure we're making real progress towards these goals in the 1.26 release.

Reply:
OK, true, the criteria don't have to be met yet, but it seems a little sketchy to claim we're planning to make GA for 1.26 when that depends on other teams doing work which isn't started yet...

Reply:
@danwinship I'm not tied to this getting to GA in 1.26, that does seem fairly unlikely at our current pace. Would you feel better if we set the GA target date as 1.27 in this PR?

Reply:
I don't know if that needs to block this - Dan, you want to make the call?

Reply:
I guess if we don't claim 1.26 now then we're blocked from doing it later? So if there's any chance it _might_ happen then I guess we want to say that...